### PR TITLE
util: fix error handling of statx()

### LINKED ISSUE #28458: Assertion 'path_is_absolute(p)' failed at src/basic/chase.c:566, function chase(). Aborting.
### systemd version the issue has been seen with

253

### Used distribution

Debian Bookworm

### Linux kernel version used

6.3.0-2

### CPU architectures issue was seen on

x86_64

### Component

_No response_

### Expected behaviour you didn't see

systemd install / upgrades without error

### Unexpected behaviour you saw

Installing new version of config file /etc/systemd/user.conf ...
Assertion 'path_is_absolute(p)' failed at src/basic/chase.c:566, function chase(). Aborting.
Aborted
Assertion 'path_is_absolute(p)' failed at src/basic/chase.c:566, function chase(). Aborting.
Aborted
dpkg: error processing package systemd (--configure):
 installed systemd package post-installation script subprocess returned error exit status 134
Errors were encountered while processing:
 systemd

### Steps to reproduce the problem

install newer systemd on debian bookwork

The error suggests there may be an issue with a path in a config file somewhere but the error isn't specific so I don't know what config file, what line, etc.

I haven't put any paths in any systemd config files anywhere.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Setting up udev (254~rc2-3) ...
Assertion 'path_is_absolute(p)' failed at src/basic/chase.c:566, function chase(). Aborting.
Aborted
Job for systemd-udevd.service failed because a fatal signal was delivered to the control process.
See "systemctl status systemd-udevd.service" and "journalctl -xeu systemd-udevd.service" for details.
invoke-rc.d: initscript udev, action "restart" failed.
● systemd-udevd.service - Rule-based Manager for Device Events and Files
     Loaded: loaded (/lib/systemd/system/systemd-udevd.service; static)
    Drop-In: /usr/lib/systemd/system/systemd-udevd.service.d
             └─syscall-architecture.conf
     Active: activating (start) since Tue 2023-07-18 21:40:42 CDT; 6ms ago
TriggeredBy: ● systemd-udevd-control.socket
             ● systemd-udevd-kernel.socket
       Docs: man:systemd-udevd.service(8)
             man:udev(7)
   Main PID: 2603017 ((md-udevd))
      Tasks: 1
     Memory: 256.0K
     CGroup: /system.slice/systemd-udevd.service
             └─2603017 "(md-udevd)"
dpkg: error processing package udev (--configure):
 installed udev package post-installation script subprocess returned error exit status 1
Processing triggers for initramfs-tools (0.142) ...
Errors were encountered while processing:
 udev
Error: Timeout was reached
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #28458.

### CODE REVIEW

File: src/basic/stat-util.c [hunk 1]
```diff
@@ -394,21 +394,35 @@ bool statx_mount_same(const struct new_statx *a, const struct new_statx *b) {
...
+static bool is_statx_fatal_error(int err, int flags) {
+        assert(err < 0);
+
+        /* If statx() is not supported or if we see EPERM (which might indicate seccomp filtering or so),
+         * let's do a fallback. Note that on EACCES we'll not fall back, since that is likely an indication of
+         * fs access issues, which we should propagate. */
+        if (ERRNO_IS_NOT_SUPPORTED(err) || err == -EPERM)
+                return false;
+
+        /* When unsupported flags specified, glibc's fallback function returns -EINVAL.
```
Comment (bluca) [inline src/basic/stat-util.c]: "are specified"

File: src/basic/mountpoint-util.c [hunk 2]
```diff
@@ -223,7 +223,7 @@ int fd_is_mount_point(int fd, const char *filename, int flags) {
                   AT_STATX_DONT_SYNC,          /* don't go to the network for this – for similar reasons */
                   STATX_TYPE,
                   &sx) < 0) {
-                if (!ERRNO_IS_NOT_SUPPORTED(errno) && !ERRNO_IS_PRIVILEGE(errno))
+                if (!ERRNO_IS_NOT_SUPPORTED(errno) && !ERRNO_IS_PRIVILEGE(errno) && errno != EINVAL)
```
Comment (bluca) [inline src/basic/mountpoint-util.c]: This deserves a comment above as it's not obvious

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
@bluca Thank you for the review. The above requests are addressed.

**Comment by mrc0mmand** [comment]
Both test-mountpoint-util and test-fd-util pass with this PR. Also, installing the just built systemd & restarting the container works fine as well, thanks!

image

Huh, haven't seen this one yet.

**Comment by yuwata** [comment]
@mrc0mmand Thank you for testing! Yay!!

**Pull request merged**